api/src/__tests__/young-inscription.test.js

```diff
@@ -19,6 +19,656 @@ jest.mock("../sendinblue", () => ({
+        parent1Status: "mother",
+        parent1FirstName: "Jane",
+        parent1LastName: "Doe",
+        parent1Email: "jane@doe.com",
+        parent1Phone: "0600020305",
+        parent2: false,
+      };
+
+      let res = await request(getAppHelper()).put("/young/inscription2023/representants/next").send(representantObj);
+      expect(res.status).toBe(200);
+      const nextResponseData = res.body.data;
+      const nextUpdatedYoung = await getYoungByIdHelper(passport.user._id);
+
+      res = await request(getAppHelper()).put("/young/inscription2023/representants/next").send(representantObj);
+      expect(res.status).toBe(200);
+      const correctionResponseData = res.body.data;
+      const correctionUpdatedYoung = await getYoungByIdHelper(passport.user._id);
+
+      delete representantObj.parent2;
+      expect(nextUpdatedYoung).toMatchObject({ ...representantObj, inscriptionStep2023: STEPS2023.DOCUMENTS });
+      expect(nextResponseData).toMatchObject({ ...representantObj, inscriptionStep2023: STEPS2023.DOCUMENTS });
+      expect(correctionUpdatedYoung).toMatchObject(representantObj);
+      expect(correctionResponseData).toMatchObject(representantObj);
+    });
+  });
+
+  describe("PUT /young/inscription2023/confirm", () => {
+    it("Should return 404 when young is not found", async () => {
+      const passport = require("passport");
+      passport.user = {};
+      let res = await request(getAppHelper()).put("/young/inscription2023/confirm");
+      expect(res.status).toBe(404);
+    });
+
+    it("Should return 200 otherwise", async () => {
+      const passport = require("passport");
+      const user = await createYoungHelper(getNewYoungFixture());
+      passport.user = user;
+
+      let res = await request(getAppHelper()).put("/young/inscription2023/confirm");
+      expect(res.status).toBe(200);
+      expect(res.body.data).toMatchObject({ informationAccuracy: "true", inscriptionStep2023: STEPS2023.WAITING_CONSENT });
+    });
+  });
+
+  describe("PUT /young/inscription2023/changeCohort", () => {
+    it("Should return 400 when no body is sent", async () => {
+      let res = await request(getAppHelper()).put("/young/inscription2023/changeCohort");
+      expect(res.status).toBe(400);
+    });
+
+    it("Should return 400 when the body sent is invalid", async () => {
+      const cohortObj = {
+        cohort: "invalid value",
+      };
+      let res = await request(getAppHelper()).put("/young/inscription2023/changeCohort").send(cohortObj);
+      expect(res.status).toBe(400);
+    });
+
+    it("Should return 404 when young is not found", async () => {
+      const passport = require("passport");
+      passport.user = {};
+      const cohortObj = {
+        cohort: "Juillet 2023",
+      };
+      let res = await request(getAppHelper()).put("/young/inscription2023/changeCohort").send(cohortObj);
+      expect(res.status).toBe(404);
+    });
+
+    // This case is not tested because the 403 status is never sent
```

Comment:
Why ? There is a canUpdateYoung function

Reply:
Je vais revoir cette partie pour être sûr. Mais dans pas mal de cas `canUpdateYoung` renvoie toujours la même chose et ne déclenche pas la 403. Je vais check à nouveau

Reply:
OK je l'ai.
La 403 n'est jamais renvoyée car il faudrait que le body contienne les éléments suivants:

```js
"status", "statusPhase1", "statusPhase2", "statusPhase3", "statusMilitaryPreparationFiles", "statusPhase2Contract"
```
La fonction `canUpdateYoung` va tester les valeurs de ces différentes propriétés pour évaluer le droit de modification.

Sauf que dans le handler de la requête, on teste la validité du `body` avec `JOI` comme ceci:

```js
const { error, value } = Joi.object({
      cohort: Joi.string().trim().valid("Février 2023 - C", "Avril 2023 - B", "Avril 2023 - A", "Juin 2023", "Juillet 2023").required(),
    }).validate(req.body, { stripUnknown: true });
```

Remarque le `{ stripUnknown: true }` à la fin. Cette configuration indique à `JOI` de supprimer toutes les clés/valeurs qui ne correspondent pas au schéma. Du coup, même si on rajoute manuellement les `status`, etc... `JOI` les supprimerai et `canUpdateYoung` renverrai `true` de toute façon.

C'est le cas pour toutes les routes de l'inscription.

Reply:
OK, ça marche. Il faudrait peut être précisé qu'il ne peut pas y avoir de 403 car on ne touche pas aux status alors !